src/Microsoft.TemplateEngine.Edge/Settings/SettingsLoader.cs

```diff
@@ -17,59 +19,102 @@
 
 namespace Microsoft.TemplateEngine.Edge.Settings
 {
-    public sealed class SettingsLoader : ISettingsLoader, IDisposable
+    internal sealed class SettingsLoader : ISettingsLoader, IDisposable
     {
-        public static readonly string HostTemplateFileConfigBaseName = ".host.json";
+        internal const string HostTemplateFileConfigBaseName = ".host.json";
         private const int MaxLoadAttempts = 20;
-        private readonly Paths _paths;
+        private static object _settingsLock = new object();
+        private static object _firstRunLock = new object();
+        private readonly SettingsFilePaths _paths;
         private readonly IEngineEnvironmentSettings _environmentSettings;
-        private SettingsStore _userSettings;
+        private readonly Action<IEngineEnvironmentSettings>? _onFirstRun;
+        private SettingsStore? _userSettings;
         private TemplateCache _userTemplateCache;
-        private IMountPointManager _mountPointManager;
-        private IComponentManager _componentManager;
-        private bool _isLoaded;
+        private IMountPointManager? _mountPointManager;
+        private IComponentManager? _componentManager;
         private bool _templatesLoaded;
         private TemplatePackageManager _templatePackagesManager;
         private volatile bool _disposed;
+        private bool _loaded;
 
-        public SettingsLoader(IEngineEnvironmentSettings environmentSettings)
+        public SettingsLoader(IEngineEnvironmentSettings environmentSettings, Action<IEngineEnvironmentSettings>? onFirstRun = null)
         {
             _environmentSettings = environmentSettings;
-            _paths = new Paths(environmentSettings);
+            _paths = new SettingsFilePaths(environmentSettings);
             _userTemplateCache = new TemplateCache(environmentSettings);
             _templatePackagesManager = new TemplatePackageManager(environmentSettings);
+            _onFirstRun = onFirstRun;
         }
 
         public IComponentManager Components
         {
             get
             {
-                EnsureLoaded();
-                return _componentManager;
+                lock (_settingsLock)
+                {
+                    if (_componentManager == null)
```

Comment:
We can first check if `_componentManager` is null. If not, we don't need to take the lock. I think this can save us some time if these methods are called frequently. We should set `_componentManager` using `Volatile.Write` method if you decide to do this.

Reply:
I'm not sure about it. Lock won't be needed if the only place when `_componentManager` can be changed is this property. However `_componentManager` can be also reset in ResetHostSettings and in bad scenario it can be reset in between operations.  I will switch to local var to avoid lock on read non null.